src/keys.rs

```diff
@@ -27,19 +38,58 @@ impl KeyPair {
     }
 
     pub fn public(&self) -> PublicKey {
-        self.public_key
+        self.public_key.clone()
+    }
+
+    pub fn sign(&self, digest: SigHash) -> Signature {
+        // TODO: Use a sensible tag
+        let ecdsa = ECDSA::from_tag(b"my-tag").enforce_low_s();
+
+        ecdsa.sign(
+            &self.secret_key.0,
+            &digest.into_inner(),
+            Derivation::Deterministic,
+        )
+    }
+
+    pub fn presign(&self, Y: PublishingPublicKey, digest: SigHash) -> EncryptedSignature {
+        // TODO: Use a sensible tag
+        let adaptor = Adaptor::from_tag(b"my-tag");
+
+        adaptor.encrypted_sign(
+            &self.secret_key.0,
+            &(Y.0).0,
+            &digest.into_inner(),
+            Derivation::Deterministic,
+        )
+    }
+}
+
+impl SecretKey {
+    pub fn public(&self) -> PublicKey {
+        let sk = &self.0;
+        // TODO: Determine what is the correct marker for this public key
+        let public_key = g!(sk * G).mark::<Normal>();
```

Comment:
It's correct. I could add something like `ECDSA::public_key_for(secret_key: &Scalar)` to avoid the need to roll your own function here.

Reply:
> I could add something like ECDSA::public_key_for(secret_key: &Scalar) to avoid the need to roll your own function here.

That would be useful! I was looking for that one lol

> This would imply having a lazy_static ECDSA or something though.

With regards to this, what do you think about what I did instantiating the `ECDSA` and `Adaptor` structs on demand? Is that how it is supposed to be used, or should I have a global instance or even one per party?

Reply:
The concern here is nonce generation and where are you going to make the decision on how to generate nonces. v0.2 secp256kfun gives you the choice whether you want to deterministic hash derivation or  "synthetic nonce" derivation which is more robust. Ideally the application provides the `NonceGen` when it instantiates the party.
If you want to instantiate it at each point then you can do:
```
let adaptor = Adaptor::<Sha256, Deterministic::<Sha256>>::default()
```
(actually you can't but let me fix that in the branch).

the short answer is that each party should have their own `Adaptor` instance. There should be a way to get a `PublicKey` from the secret key via the instance to perhaps remove the need for some conversions here.

Reply:
There is `Adaptor::encryption_key_for` and `ECDSA::verification_key_for`. See https://github.com/LLFourn/secp256kfun/pull/16/commits/788e676533ac73b95e868a21a45827a301ff08c4